### PR TITLE
Introduce PUNCH_MY_PATH frame for multipath

### LINKED ISSUE #30: Compatibility with multipath
The proposed mechanism is to have server and clients start "path validation" simultaneously, using the "PUNCH_ME_NOW" frame for synchronization. I assume that the goal is to create a single path.

There is no problem doing that with RFC 9000, but there is an issue if the multipath extension is negotiated, because then path creation implies picking a path ID -- i.e., both server and client pick a CID associated with the same path-ID. If they don't, we get two paths.

The simplest way to do that would be to add a "path_id" element in the PUNCH_ME_NOW frame.

### PR DESCRIPTION
**Pull request by huitema**
I looked at what if would take to support multipath. I used a new frame for "PUNCH_ME_NOW + Path ID", but if we could have a bit in the PUNCH_ME_NOW frame type we would not need that. We would need some explanation about how to use the Path ID, which I added in a subsection of PUNCH_ME_NOW.

I expect this text to evolve after discussions. But I would like to have these discussions.

Close #30

### CODE REVIEW

File: draft-seemann-quic-nat-traversal.md [hunk 1]
```diff
@@ -178,6 +178,39 @@ number of concurrent path validations.
...
 
+### Probing Paths when using QUIC Multipath
+
+If the QUIC Multipath extensions have been negotiated (see {{MULTIPATH}}),
+the probing of paths will be as described in {{probing-paths}}, but the
+client should send the candidate pairs using PUNCH_MY_PATH frames instead
+of PUNCH_ME_NOW frames. The PUNCH_MY_PATH frames carry the same information
+as the PUNCH_ME_NOW frame, plus a path identifier (Path ID), chosen
+from the range of valid Path ID. The server should
+use connection identifiers associated to that Path-ID when validating
```
Comment (ekinnear) [inline draft-seemann-quic-nat-traversal.md]: Silly nit, is it `Path ID` or `Path-ID`?

File: draft-seemann-quic-nat-traversal.md [hunk 2]
```diff
@@ -289,6 +322,41 @@ This frame is only sent from the client to the server. Clients MUST treat
...
+    Port (16),
+}
+~~~
+
+The PUNCH_MY_PATH frame is identical to the PUNCH_ME_NOW
+frame, except for the addition of the Path-Id element:
+
+Path ID:
+
+: The path Id selected by the client for the new path.
```
Comment (ekinnear) [inline draft-seemann-quic-nat-traversal.md]:
```suggestion
: The Path ID selected by the client for the new path.
```

File: draft-seemann-quic-nat-traversal.md [hunk 3]
```diff
@@ -289,6 +322,41 @@ This frame is only sent from the client to the server. Clients MUST treat
...
+valid if the QUIC Extension for Multipath has been negotiated,
+(see {{quic-multipath}}. Clients MUST treat
+receipt of a PUNCH_MY_PATH frame as a connection error of type
+PROTOCOL_VIOLATION.
+
+*TODO* There is lots of duplicated text here. This could be
+eliminated if we allocate a range of 4 values for the PUNCH_ME_NOW
+frame ID, with the least significant bit indicating IPv4/IPv6,
+and the second least significant bit indicating absence/presence
+of the Path ID.
```
Comment (ekinnear) [inline draft-seemann-quic-nat-traversal.md]: It seems reasonable to have the frame have a flag indicating presence of the single additional field. We don't seem them otherwise diverging much in the future, right?

File: draft-seemann-quic-nat-traversal.md [hunk 4]
```diff
@@ -178,6 +178,39 @@ number of concurrent path validations.
...
+Different PUNCH_MY_PATH frames may be sent with different path identifiers,
+which may result in the creation or migration of as many paths.
+
+If the server receives a PUNCH_ME_NOW frame after the QUIC Multipath extensions
+have been negotiated, it MUST treat it as PUNCH_MY_PATH frame carrying Path ID 0.
+
+When QUIC Multipath extensions is been negotiated, the probing of new
+path will be limited by the availability of connection identifiers
+for new Path IDs. Endpoints SHOULD set Path ID limits high enough to allow
+for the desired number of concurrent path validation attempts.
```
Comment (ekinnear) [inline draft-seemann-quic-nat-traversal.md:209]: No objection to helping people do the right thing here, but for my understanding -- this is true of all multipath, no? And we'd note it here because we might be asking people to try more paths than they otherwise would without NAT traversal?

File: draft-seemann-quic-nat-traversal.md [hunk 5]
```diff
@@ -178,6 +178,39 @@ number of concurrent path validations.
...
+
+If the selected Path ID is not yet in use, successful path probing will
+result in the creation of a new path, to be used in combination with the
+pre-existing paths.
+
+If the selected Path ID is already in use, successful path probing will
+result in the migration of the existing path to a new address.
+
+Different PUNCH_MY_PATH frames may be sent with different path identifiers,
+which may result in the creation or migration of as many paths.
```
Comment (ekinnear) [inline draft-seemann-quic-nat-traversal.md]: Wording nit: We can probably reword this to make it more clear, something like you'd have "as many paths as PUNCH_MY_PATH frames" or similar.

File: draft-seemann-quic-nat-traversal.md [hunk 6]
```diff
@@ -178,6 +178,39 @@ number of concurrent path validations.
...
+
+If the server receives a PUNCH_ME_NOW frame after the QUIC Multipath extensions
+have been negotiated, it MUST treat it as PUNCH_MY_PATH frame carrying Path ID 0.
+
+When QUIC Multipath extensions is been negotiated, the probing of new
```
Comment (ekinnear) [inline draft-seemann-quic-nat-traversal.md]:
```suggestion
When the QUIC Multipath extensions have been negotiated, the probing of new
```

File: draft-seemann-quic-nat-traversal.md [hunk 7]
```diff
@@ -178,6 +178,39 @@ number of concurrent path validations.
...
+from the range of valid Path ID. The server should
+use connection identifiers associated to that Path-ID when validating
+the path.
+
+If the selected Path ID is not yet in use, successful path probing will
+result in the creation of a new path, to be used in combination with the
+pre-existing paths.
+
+If the selected Path ID is already in use, successful path probing will
+result in the migration of the existing path to a new address.
```
Comment (ekinnear) [inline draft-seemann-quic-nat-traversal.md]: Just to make sure I understand, we're saying: 
- If this is a new path, it just brings up a new path, potentially having hole-punched it
- If it's an existing path, please hole-punch it over here, like if we did PUNCH_ME_NOW, but since it's MP-QUIC, just move that path and not everything else

**Review by ekinnear — COMMENTED** [review]
Probably worth another editorial pass as you noted, but overall seems like a straightforward mapping of the core concepts to multipath -- nice!

File: draft-seemann-quic-nat-traversal.md (hunk 1 above)
Comment (huitema) [reply draft-seemann-quic-nat-traversal.md]: The multipath draft uses "path ID" throughout.

File: draft-seemann-quic-nat-traversal.md (hunk 4 above)
Comment (huitema) [reply draft-seemann-quic-nat-traversal.md:209]: Yes. Same reason as the blurb on connection ID limits

**Comment by huitema** [comment]
Based on @ekinnear comments, I think I should do the clean edit now (tomorrow?), unify PUNCH_MY_PATH and PUNCH_ME_NOW, etc. That would be cleaner.

File: draft-seemann-quic-nat-traversal.md (hunk 3 above)
Comment (huitema) [reply draft-seemann-quic-nat-traversal.md]: Yes. That means changing the current provisional allocation of the frame type for PUNCH_ME_NOW.

File: draft-seemann-quic-nat-traversal.md (hunk 7 above)
Comment (huitema) [reply draft-seemann-quic-nat-traversal.md]: I rewrote that part, because the Multipath logic will not normally do path migration. Having a packet arriving with a new tuple on an existing path signals a NAT rebinding. That will trigger a Path Challenge on old and new path, keeping the old path if it is still available, maybe migrating to the new path if it responds sooner. Overall, picking an unused path ID is much cleaner. Even if the desired result is path migration, using a new path ID provides "make before break" functionality.

**Comment by huitema** [comment]
This PR would close the "compatibility with multipath" issue.

Close #30
